### PR TITLE
functions writing bambu outputs to gtf and txt files and creating a GRangesList from GTF file

### PR DESCRIPTION
**Pull request by yuukiiwa**
There are three functions:
write.gtf reads in a GRangesList and an optional geneID (column 2) dataframe with their corresponding transcriptIDs (column 1).
write.bambu takes in a SummarizedExperiment object from bambu and outputs a gtf file by running write.gtf and counts files for transcripts and genes.
read.gtf reads in a gtf file and outputs a GRangesList, where eqClass is not included.
Should I just use TXNAME as the eqClass?
Thanks.

### CODE REVIEW

File: R/readWrite.R [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#' @title transcript to gene expression
+#' @param se a summarizedExperiment object from \code{\link{bambu}}
+#' @param path the destination of the output files (gtf, transcript counts, and gene counts)
+#' @return gtf a GTF dataframe
+#' @export
+write.bambu <- function(transcript_se,path){
+  if (missing(transcript_se) | missing(path)){
+    stop('Both summarizedExperiment object from bambu and the path for the output files are required.')
+  }else{
+    dir.create(path)
```
Comment (jonathangoeke) [inline R/readWrite.R:11]: test if the directory exists before creating it

File: R/readWrite.R [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }else{
+      gtf_trns <- data.frame(seqname=df$seqnames, source= "Bambu",feature= "transcript",
+                             start=df$start,end=df$uend,score=".",strand=df$strand,frame=".",
+                             attributes= paste(df$GENEID,df$group_name))
+    }
+    gtf <- rbind(gtf_trns,gtf_exon)
+    gtf <- gtf[order(gtf$attributes),]
+  } 
+  write.table(gtf, file= file, quote=FALSE, row.names=FALSE, col.names=FALSE, sep = "\t")
+  return (gtf)
```
Comment (jonathangoeke) [inline R/readWrite.R:72]: the function does not need to return the gtf object

File: R/readWrite.R [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+#' Outputs GRangesList object from reading a GTF file
+#' @title transcript to gene expression
+#' @param gtf_file a GTF file
+#' @return grlist a GRangesList object
+#' @export
+read.gtf <- function(gtf_file){
```
Comment (jonathangoeke) [inline R/readWrite.R:79]: the parameter can just be file instead of gtf_file

File: R/readWrite.R [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  return (gtf)
+}
+#' Outputs a GTF file for the nanorna-bam nextflow pipeline 
+#' @title transcript to gene expression
+#' @param grList a GRangesList object
+#' @param file the output gtf file name
+#' @param geneIDs an optional dataframe of geneIDs (column 2) with the corresponding transcriptIDs (column 1)
+#' @return gtf a GTF dataframe
+#' @export
+write.gtf <- function (grList,file,geneIDs) {
```
Comment (jonathangoeke) [inline R/readWrite.R:33]: to be consistent with bambu, the parameters could be (annotation, file, geneIDs = NULL). When annotation was created the geneIDs are optional as they are stored in mcols(annotation)$GENEID (can be tested with !is.null(mcols(annotation, use.names=FALSE)$GENEID))

File: R/readWrite.R [hunk 5]
```diff
@@ -0,0 +1,96 @@
+#' Outputs a GTF file, transcript-count file, and gene-count file from bambu
+#' @title transcript to gene expression
+#' @param se a summarizedExperiment object from \code{\link{bambu}}
+#' @param path the destination of the output files (gtf, transcript counts, and gene counts)
+#' @return gtf a GTF dataframe
+#' @export
+write.bambu <- function(transcript_se,path){
```
Comment (jonathangoeke) [inline R/readWrite.R:7]: paramters could be names (se, path)

File: R/readWrite.R [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#' Outputs a GTF file, transcript-count file, and gene-count file from bambu
+#' @title transcript to gene expression
```
[the remark this answers is not in the document]
Comment (yuukiiwa) [reply R/readWrite.R]: I have changed the titles for all the functions.

File: R/readWrite.R (hunk 5 above)
Comment (yuukiiwa) [reply R/readWrite.R]: The parameter transcript_se has been changed to se.

File: R/readWrite.R (hunk 3 above)
Comment (yuukiiwa) [reply R/readWrite.R]: The parameter gtf_file has been changed to file.

File: R/readWrite.R (hunk 2 above)
Comment (yuukiiwa) [reply R/readWrite.R]: The function does not return a GTF dataframe now. I have also removed the test_that for the GTF file format.

File: R/readWrite.R [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    gtf <- write.gtf(grList=transcript_grList,file=transcript_gtffn,geneIDs=transcript_geneIDs)
+    transcript_counts <- as.data.frame(assays(transcript_se)$counts)
+    transcript_countsfn <- paste(path,"counts_transcript.txt",sep="/")
+    write.table(transcript_counts, file= transcript_countsfn, sep="\t",quote=FALSE)
+    gene_se <- transcriptToGeneExpression(se)
+    gene_counts <- as.data.frame(assays(gene_se)$counts)
+    gene_countsfn <- paste(path,"counts_gene.txt",sep="/")
+    write.table(gene_counts, file= gene_countsfn, sep="\t", quote=FALSE)
+    }
+  return (gtf)
```
[the remark this answers is not in the document]
Comment (yuukiiwa) [reply R/readWrite.R]: The function does not return a GTF dataframe now. I have also removed the test_that for the GTF file format.

File: R/readWrite.R [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+#' @param path the destination of the output files (gtf, transcript counts, and gene counts)
+#' @return gtf a GTF dataframe
+#' @export
+write.bambu <- function(transcript_se,path){
+  if (missing(transcript_se) | missing(path)){
+    stop('Both summarizedExperiment object from bambu and the path for the output files are required.')
+  }else{
+    dir.create(path)
+    transcript_grList <- rowRanges(transcript_se)
+    transcript_gtffn <- paste(path,"bambu_transcript_exon.gtf",sep="/")
```
[the remark this answers is not in the document]
Comment (yuukiiwa) [reply R/readWrite.R]: The path can now have a prefix.

File: R/readWrite.R (hunk 1 above)
Comment (yuukiiwa) [reply R/readWrite.R]: The output directory will only be created if it has not been existing.

File: R/readWrite.R (hunk 4 above)
Comment (yuukiiwa) [reply R/readWrite.R]: The parameters are now (annotation, file, geneIDs = NULL). Now the write.gtf function will output gtf files with both gene and transcript IDs even if geneIDs  = NULL. The write.bambu's write.gtf will also not be taking geneIDs to generate a gtf file.

**Comment by yuukiiwa** [comment]
I have added in the prepareAnnotationFromGTF_draft function into the readWrite.R file. It can output a GRangesList that can run through bambu and generate a SummarizedExperiment object. However, it encountered problems in the transcriptToGeneExpression function (line 38, 52, and 65), where the RowNames in exByGene (generated by txRangesToGeneRanges) are the merged of GENEID and TXNAME, and to have the transcriptToGeneExpression work, it requires the exByGene RowNames to be GENEID alone. 
Thanks.

**Comment by jonathangoeke** [comment]
Thanks!

**Pull request merged**